Update docs for the knapsack_pro gem and related articles (#152)
KnapsackPro/docs.knapsackpro.com: docusaurus/docs/ruby/rspec.mdx

<Tabs>
  <TabItem value="junit-formatter" label="JUnit XML" default>
    ```bash
    bundle exec rake "knapsack_pro:rspec[--format documentation --format RspecJunitFormatter --out tmp/rspec.xml]"
    ```
  </TabItem>
  <TabItem value="json-formatter" label="JSON">
    ```bash
    bundle exec rake "knapsack_pro:rspec[--format documentation --format json --out tmp/rspec.json]"
    ```
  </TabItem>
</Tabs>

### Queue Mode

You need to install the [`rspec_junit_formatter`](https://github.com/sj26/rspec_junit_formatter) gem.

<Tabs>
  <TabItem value="junit-formatter" label="JUnit XML" default>
    ```bash
    bundle exec rake "knapsack_pro:queue:rspec[--format documentation --format RspecJunitFormatter --out tmp/rspec.xml]"
    ```

    You need to append the CI node index to the report name to avoid conflicts if your CI nodes write to the same disk:

    ```bash
    # Refer to your CI docs for `$MY_CI_NODE_INDEX`
    export KNAPSACK_PRO_CI_NODE_INDEX=$MY_CI_NODE_INDEX

    bundle exec rake "knapsack_pro:queue:rspec[--format documentation --format RspecJunitFormatter --out tmp/rspec_$KNAPSACK_PRO_CI_NODE_INDEX.xml]"
    ```

>     This applies also if you are running parallel test processes on each CI node (see our page on to integrate Knapsack Pro with [`parallel_tests`](parallel_tests.md) for an example).

Reviewer: ```suggestion
    This applies also if you are running parallel test processes on each CI node (see how to [integrate Knapsack Pro with `parallel_tests`](parallel_tests.md)).
```

>     This applies also if you are running parallel test processes on each CI node (see our page on to integrate Knapsack Pro with [`parallel_tests`](parallel_tests.md) for an example).

Reviewer: Isn't this also true for Regular? Can we merge the Regular and Queue sections into one?

Maybe we could have separate sections for JUnit and JSON and use tabs for Queue and Regular (with Regular second).

I'd like to hide Regular vs Queue when possible.

Author: > Maybe we could have separate sections for JUnit and JSON and use tabs for Queue and Regular (with Regular second).
> 
> I'd like to hide Regular vs Queue when possible.

Good idea. Done: https://github.com/KnapsackPro/docs.knapsackpro.com/pull/152/commits/23f72c454c219b8625063908d6d3d193b55272d7

Author: > Isn't this also true for Regular? Can we merge the Regular and Queue sections into one?

We should not merge this. Additionally, recommending the use of Regular Mode with parallel_tests is not advisable.

Technically, you can use Regular Mode with the parallel_tests gem, but it's not practical to recommend this approach. For even distribution of tests with parallel_tests, a dynamic test split (Queue Mode) is necessary. Queue Mode is simply superior.

  </TabItem>
  <TabItem value="json-formatter" label="JSON">
    ```bash
    bundle exec rake "knapsack_pro:queue:rspec[--format documentation --format json --out tmp/rspec.json]"
    ```

    You need to append the CI node index to the report name to avoid conflicts if your CI nodes write to the same disk:

    ```bash
    # Refer to your CI docs for `$MY_CI_NODE_INDEX`
    export KNAPSACK_PRO_CI_NODE_INDEX=$MY_CI_NODE_INDEX

    bundle exec rake "knapsack_pro:queue:rspec[--format documentation --format json --out tmp/rspec_$KNAPSACK_PRO_CI_NODE_INDEX.json]"
    ```

>     This applies also if you are running parallel test processes on each CI node (see our page on to integrate Knapsack Pro with [`parallel_tests`](parallel_tests.md) for an example).

Reviewer: ```suggestion
    This applies also if you are running parallel test processes on each CI node (see how to [integrate Knapsack Pro with `parallel_tests`](parallel_tests.md)).
```

bundle exec rake "knapsack_pro:queue:rspec"
```

> ## Load code only once for a specific type of specs using `when_first_matching_example_defined`

Reviewer: ```suggestion
## Load code only once for a specific type of specs
```

Reviewer: Did anybody request this? I'd like to see how they phrased it. I'd like to change the copy here a bit.

Author: Yes. Please see: https://knapsackpro.slack.com/archives/C05RJKCQ42D/p1706891273447509?thread_ts=1702626494.818529&cid=C05RJKCQ42D

bundle exec rake "knapsack_pro:queue:rspec"
```

## Load code only once for a specific type of specs

> You can save time and run tests faster if you avoid loading a service that is not needed to run tests.
> 
> For example, your system tests require `MyExpensiveService` to run. Tests are allocated dynamically between parallel CI nodes in Queue Mode. A CI node may not have a system test assigned to it. There is no point in booting `MyExpensiveService` in such a case.
> 
> You can configure RSpec to load the service dependency only when needed. It is loaded only once after the first system test is fetched from the Queue API.
> 
> Please add the following helper to your `spec_helper.rb`:

Reviewer: ```suggestion
In large projects that use RSpec, it’s common to have some expensive setup logic that is only needed when certain kinds of specs have been loaded. If that kind of spec has not been loaded, you’d prefer to avoid the cost of doing the setup.

The `when_first_matching_example_defined` hook makes it easy to conditionally perform some logic when the first example is defined with matching metadata, allowing you to ensure the necessary setup is performed only when needed.

Due to the internals of Knapsack Pro, you need to add the following code to `spec_helper.rb` to use `when_first_matching_example_defined` with Queue Mode:
```

### `before(:suite)` / `after(:suite)` are executed multiple times in Queue Mode

> Update the `knapsack_pro` gem to `>= 7.0` in order to run them only once. Learn more about [hooks](hooks.mdx).

Reviewer: ```suggestion
This was fixed in `knapsack_pro`  v7.0.
```

### `before(:suite)` / `after(:suite)` are executed multiple times in Queue Mode

> This was fixed in `knapsack_pro`  v7.0.

Reviewer: Sorry, I know that I already asked you to change it once already.

```suggestion
This issue has been fixed in knapsack_pro version 7.0. Learn more about [Knapsack Pro hooks](https://docs.knapsackpro.com/ruby/hooks/).
```
